signing/signing.go

```diff
@@ -1,20 +1,363 @@
+	}
+
+	cert := certs[0]
+
+	// Verify that cert isn't expired.
+	if err = cosign.CheckExpiry(cert, time.Unix(*entry.IntegratedTime, 0)); err != nil {
+		return "", "", "", "", fmt.Errorf("cosign certificate is expired: %v", err)
+	}
+
+	// Get repo reference & path from cert.
+	for _, ext := range cert.Extensions {
+		// OID source: https://github.com/sigstore/fulcio/blob/96ef49cc7662912ba37d46f738757e8d8d5b5355/docs/oid-info.md#L33
+		// TODO: retrieve these by name.
+		if ext.Id.String() == "1.3.6.1.4.1.57264.1.6" {
+			repoRef = string(ext.Value)
+		}
+		if ext.Id.String() == "1.3.6.1.4.1.57264.1.5" {
+			repoPath = string(ext.Value)
+		}
+		if ext.Id.String() == "1.3.6.1.4.1.57264.1.3" {
+			repoSHA = string(ext.Value)
+		}
+	}
+
+	if len(repoRef) == 0 || len(repoPath) == 0 {
+		return "", "", "", "", fmt.Errorf("error extracting repo ref or path from certificate %v", err)
+	}
+
+	// Get workflow job ref from the certificate.
+	certURIs := cert.URIs
+	if len(certURIs) == 0 {
+		return "", "", "", "", errors.New("error: certificate has no URIs")
+	}
+	workflowRef := certURIs[0].Path
+	if len(workflowRef) == 0 {
+		return "", "", "", "", errors.New("error: workflow path is empty")
+	}
+
+	// Remove repo path from workflow filepath.
+	ind := strings.Index(workflowRef, repoPath) + len(repoPath) + 1
+	workflowPath = workflowRef[ind:]
+
+	// Remove repo ref tag from workflow filepath.
+	workflowPath = workflowPath[:strings.Index(workflowPath, "@")]
+
+	return repoPath, repoRef, repoSHA, workflowPath, nil
+}
+
+func verifyScorecardWorkflow(workflowContent string) error {
+	// Verify workflow contents using actionlint.
+	workflow, lintErrs := actionlint.Parse([]byte(workflowContent))
+	if lintErrs != nil {
+		return fmt.Errorf("actionlint errors parsing workflow: %v", lintErrs)
+	}
+
+	// Verify that there are no global env vars or defaults.
+	if workflow.Env != nil || workflow.Defaults != nil {
+		return errors.New("workflow contains global env vars or defaults")
+	}
+
+	// Verify that the only global permission set is read-all.
+	if workflow.Permissions.All.Value != "read-all" {
+		return errors.New("workflow permission isn't read-all")
+	}
+
+	// Extract main (analysis) job
+	jobsMap := workflow.Jobs
+	jobs := reflect.ValueOf(jobsMap).MapKeys()
+
+	if len(jobs) != 1 {
+		return errors.New("number of jobs isn't 1")
+	}
+	job := jobsMap[jobs[0].String()]
+	if job == nil {
+		return errors.New("workflow doens't have a main job")
+	}
+
+	// Verify that there is no job container or services.
+	if job.Container != nil || len(job.Services) > 0 {
+		return errors.New("workflow contains container or service")
+	}
+
+	// Verify that the workflow runs on ubuntu-latest and nothing else.
+	if job.RunsOn == nil {
+		return errors.New("no RunsOn found in workflow")
+	} else {
+		labels := job.RunsOn.Labels
+		if len(labels) == 0 || len(labels) > 1 || labels[0].Value != "ubuntu-latest" {
```

Comment:
we can accept other flavor of ubuntu. GitHub has a page with a list of Ubuntu. Not super important, but you never know :-)